File: examples/cpp/round_trip_example.cpp
#include <algorithm>
#include <cstdint>
#include <cstring>
#include <iomanip>
#include <iostream>
#include <memory>
#include <sstream>
#include <string>

#include "client/cpp/payload_manager_client.h"
#include "otel_tracer.hpp"
#include "payload/manager/v1.hpp"
#include "traced_channel.hpp"

namespace {

std::string UuidToHex(const std::string& uuid_bytes) {
  std::ostringstream os;
  for (size_t i = 0; i < uuid_bytes.size(); ++i) {
    if (i == 4 || i == 6 || i == 8 || i == 10) {
      os << '-';
    }
    os << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(static_cast<unsigned char>(uuid_bytes[i]));
  }
  return os.str();
}

} // namespace

int main(int argc, char** argv) {
  // argv[1]: server endpoint  (default localhost:50051)
  // argv[2]: OTLP gRPC endpoint (default localhost:4317, empty to disable)
  const std::string target  = argc > 1 ? argv[1] : "localhost:50051";
  const std::string otlp_ep = argc > 2 ? argv[2] : "localhost:4317";

  OtelInit(otlp_ep, "cpp-examples");
  auto channel = StartSpanAndMakeChannel(target, "round_trip_example");
  payload::manager::client::PayloadClient client(channel);

  // Allocate a writable payload in RAM, fill it locally, then commit so it
  // becomes visible to readers.
  constexpr uint64_t kPayloadSize = 64;
  auto               writable     = client.AllocateWritableBuffer(kPayloadSize, payload::manager::v1::TIER_RAM);
  if (!writable.ok()) {
    std::cerr << "AllocateWritableBuffer failed: " << writable.status().ToString() << '\n';
    return 1;
  }

  auto writable_payload = writable.ValueOrDie();
  for (uint64_t i = 0; i < kPayloadSize; ++i) {
    writable_payload.buffer->mutable_data()[i] = static_cast<uint8_t>(i & 0xFFu);
  }

  const auto& payload_id    = writable_payload.descriptor.payload_id();
  const auto  uuid_text     = UuidToHex(payload_id.value());
  auto        commit_status = client.CommitPayload(payload_id);
  if (!commit_status.ok()) {
    std::cerr << "CommitPayload failed: " << commit_status.ToString() << '\n';
    return 1;
  }

  // Acquire a read lease to validate what was committed.
  auto readable = client.AcquireReadableBuffer(payload_id);
  if (!readable.ok()) {
    std::cerr << "AcquireReadableBuffer failed: " << readable.status().ToString() << '\n';
    return 1;
  }

  auto readable_payload = readable.ValueOrDie();
  std::cout << "Committed and acquired payload UUID=" << uuid_text << ", size=" << readable_payload.buffer->size() << " bytes\n";

  const int preview_len = static_cast<int>(std::min<int64_t>(8, readable_payload.buffer->size()));
  std::cout << "First " << preview_len << " bytes:";
  for (int i = 0; i < preview_len; ++i) {
    std::cout << ' ' << static_cast<int>(readable_payload.buffer->data()[i]);
  }
  std::cout << '\n';

  // Release the lease to avoid holding resources/pinning placement.
  auto release_status = client.Release(readable_payload.lease_id);
  if (!release_status.ok()) {
    std::cerr << "Release failed: " << release_status.ToString() << '\n';
    return 1;
  }

  OtelEndSpan();
  OtelShutdown();
  return 0;
}

File: examples/python/round_trip_example.py
#!/usr/bin/env python3

import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(REPO_ROOT / "client/python"))
import uuid

import grpc

from payload_manager_client import PayloadClient
from payload.manager.core.v1 import placement_pb2


def main() -> int:
    # Allow overriding the gRPC endpoint from the command line so this script can
    # run against local, containerized, or remote Payload Manager deployments.
    target = sys.argv[1] if len(sys.argv) > 1 else "localhost:50051"
    client = PayloadClient(grpc.insecure_channel(target))

    payload_size = 64
    # Allocate a writable RAM-tier buffer. The server returns a descriptor
    # (including UUID) plus an mmap-backed byte region we can fill directly.
    writable = client.AllocateWritableBuffer(payload_size, placement_pb2.TIER_RAM)
    writable.mmap_obj[:] = bytes((i & 0xFF) for i in range(payload_size))

    # Use the canonical PayloadID protobuf returned by AllocatePayload and
    # derive a printable UUID only for logging.
    payload_id = writable.descriptor.id
    payload_uuid = uuid.UUID(bytes=bytes(payload_id.value))
    # Commit makes the payload visible/immutable for readers.
    client.CommitPayload(payload_id)

    # AcquireReadableBuffer returns a lease-scoped view of committed data.
    readable = client.AcquireReadableBuffer(payload_id)
    print(f"Committed and acquired payload UUID={payload_uuid}, size={readable.buffer.size} bytes")

    preview_len = min(8, readable.buffer.size)
    preview = " ".join(str(x) for x in readable.buffer.to_pybytes()[:preview_len])
    print(f"First {preview_len} bytes: {preview}")

    # Always release leases once done to avoid pinning payload placement.
    client.Release(readable.lease_id)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
